File: brain.js
// ============================================================
// JARVIS 3.2
// brain.js
// Virtual Brain
// ============================================================

class JarvisBrain {

    constructor() {
        this.name = "JARVIS";
        this.history = [];
    }

    // --------------------------------------------------------
    // MEMORY
    // --------------------------------------------------------

    remember(userText, jarvisText) {

        this.history.push({
            user: userText,
            jarvis: jarvisText,
            time: new Date().toISOString()
        });

        // Keep the browser session lightweight
        if (this.history.length > 20) {
            this.history.shift();
        }
    }


    // --------------------------------------------------------
    // UNDERSTAND COMMAND
    // --------------------------------------------------------

    understand(text) {

        const original = text.trim();
        const lower = original.toLowerCase();


        // =========================
        // YOUTUBE
        // =========================

        if (lower === "open youtube" ||
            lower.includes("open youtube")) {

            return {
                type: "command",
                command: "open_youtube",
                argument: ""
            };
        }


        if (
            lower.includes("search youtube for") ||
            lower.includes("search youtube")
        ) {

            let query = lower
                .replace("search youtube for", "")
                .replace("search youtube", "")
                .trim();

            return {
                type: "command",
                command: "youtube_search",
                argument: query
            };
        }


        // =========================
        // GOOGLE
        // =========================

        if (
            lower === "open google" ||
            lower.includes("open google")
        ) {

            return {
                type: "command",
                command: "open_google",
                argument: ""
            };
        }


        if (lower.includes("search google for")) {

            let query = lower
                .replace("search google for", "")
                .trim();

            return {
                type: "command",
                command: "google_search",
                argument: query
            };
        }


        if (lower.startsWith("search for")) {

            let query = lower
                .replace("search for", "")
                .trim();

            return {
                type: "command",
                command: "google_search",
                argument: query
            };
        }


        // =========================
        // GMAIL
        // =========================

        if (
            lower === "open gmail" ||
            lower.includes("open gmail")
        ) {

            return {
                type: "command",
                command: "open_gmail",
                argument: ""
            };
        }


        if (lower.includes("search gmail for")) {

            let query = lower
                .replace("search gmail for", "")
                .trim();

            return {
                type: "command",
                command: "gmail_search",
                argument: query
            };
        }


        // =========================
        // WHATSAPP
        // =========================

        if (
            lower === "open whatsapp" ||
            lower.includes("open whatsapp")
        ) {

            return {
                type: "command",
                command: "open_whatsapp",
                argument: ""
            };
        }


        // =========================
        // TIME
        // =========================

        if (
            lower === "time" ||
            lower.includes("what time") ||
            lower.includes("current time")
        ) {

            return {
                type: "command",
                command: "get_time",
                argument: ""
            };
        }


        // =========================
        // DATE
        // =========================

        if (
            lower === "date" ||
            lower.includes("what date") ||
            lower.includes("today's date")
        ) {

            return {
                type: "command",
                command: "get_date",
                argument: ""
            };
        }


        // =========================
        // CALL
        // =========================

        if (lower.startsWith("call ")) {

            const contact = original
                .substring(5)
                .trim();

            return {
                type: "command",
                command: "call",
                argument: contact
            };
        }


        // =========================
        // CONVERSATION
        // =========================

        return {
            type: "conversation",
            command: null,
            argument: ""
        };
    }


    // --------------------------------------------------------
    // NORMAL AI-STYLE RESPONSE
    // --------------------------------------------------------

    generateResponse(text) {

        const lower = text.toLowerCase().trim();


        if (
            lower === "hello" ||
            lower === "hi" ||
            lower === "hey"
        ) {

            return "Hello. I'm Jarvis. How can I help you?";
        }


        if (lower.includes("how are you")) {

            return "I'm online and ready to help.";
        }


        if (lower.includes("who are you")) {

            return "I am Jarvis, your personal AI assistant.";
        }


        if (
            lower.includes("thank you") ||
            lower.includes("thanks")
        ) {

            return "You're welcome.";
        }


        if (lower.includes("good morning")) {

            return "Good morning. Jarvis is online.";
        }


        if (lower.includes("good night")) {

            return "Good night. I'll be here when you need me.";
        }


        if (lower.includes("jarvis")) {

            return "Yes, I'm listening.";
        }


        return (
            "I understand you. " +
            "I'm ready to connect to my generative AI system."
        );
    }


    // --------------------------------------------------------
    // PROCESS MESSAGE
    // --------------------------------------------------------

    async process(text) {

        const result = this.understand(text);


        // =========================
        // COMMAND
        // =========================

        if (result.type === "command") {

            let response = "";

            switch (result.command) {

                case "open_youtube":

                    response = "Opening YouTube.";
                    break;


                case "youtube_search":

                    response =
                        "Searching YouTube for " +
                        result.argument + ".";

                    break;


                case "open_google":

                    response = "Opening Google.";
                    break;


                case "google_search":

                    response =
                        "Searching Google for " +
                        result.argument + ".";

                    break;


                case "open_gmail":

                    response = "Opening Gmail.";
                    break;


                case "gmail_search":

                    response =
                        "Searching Gmail for " +
                        result.argument + ".";

                    break;


                case "open_whatsapp":

                    response = "Opening WhatsApp.";
                    break;


                case "get_time":

                    response =
                        "The time is " +
                        new Date().toLocaleTimeString();

                    break;


                case "get_date":

                    response =
                        "Today is " +
                        new Date().toLocaleDateString();

                    break;


                case "call":

                    response =
                        "I received your request to call " +
                        result.argument +
                        ". Phone integration will be connected later.";

                    break;


                default:

                    response = "Command received.";
            }


            // Send command to commands.js
            if (
                typeof executeCommand === "function"
            ) {

                try {

                    await executeCommand(
                        result.command,
                        result.argument
                    );

                } catch (error) {

                    console.error(
                        "Command error:",
                        error
                    );
                }
            }


            this.remember(text, response);

            return response;
        }


        // =========================
        // GENERATIVE AI RESPONSE
        // =========================

        const response =
            this.generateResponse(text);

        this.remember(text, response);

        return response;
    }
}


// ============================================================
// CREATE GLOBAL JARVIS BRAIN
// ============================================================

window.jarvisBrain = new JarvisBrain();

File: brain.py
"""
JARVIS 3.2
Virtual Brain
"""

from commands import execute_command


# ============================================================
# JARVIS BRAIN
# ============================================================

class JarvisBrain:

    def __init__(self):
        self.name = "JARVIS"
        self.history = []

    # --------------------------------------------------------
    # Remember conversation during the current session
    # --------------------------------------------------------

    def remember(self, user_text, jarvis_text):
        self.history.append({
            "user": user_text,
            "jarvis": jarvis_text
        })

        # Keep the session lightweight
        if len(self.history) > 20:
            self.history.pop(0)

    # --------------------------------------------------------
    # Detect simple commands
    # --------------------------------------------------------

    def understand(self, text):

        text = text.lower().strip()

        # YouTube
        if "open youtube" in text:
            return "open_youtube", ""

        if "youtube" in text and (
            "search" in text or
            "find" in text or
            "look for" in text
        ):
            query = text

            for phrase in [
                "search youtube for",
                "search youtube",
                "find on youtube",
                "look for on youtube"
            ]:
                query = query.replace(phrase, "")

            return "youtube_search", query.strip()

        # Google
        if "open google" in text:
            return "open_google", ""

        if "search google for" in text:
            query = text.replace(
                "search google for", ""
            ).strip()

            return "google_search", query

        if text.startswith("search for "):
            query = text.replace(
                "search for", ""
            ).strip()

            return "google_search", query

        # Gmail
        if "open gmail" in text:
            return "open_gmail", ""

        if "search gmail for" in text:
            query = text.replace(
                "search gmail for", ""
            ).strip()

            return "gmail_search", query

        # WhatsApp
        if "open whatsapp" in text:
            return "open_whatsapp", ""

        # Time
        if "what time" in text or text == "time":
            return "get_time", ""

        # Date
        if "what date" in text or text == "date":
            return "get_date", ""

        # Call
        if text.startswith("call "):
            contact = text.replace(
                "call", "", 1
            ).strip()

            return "call", contact

        return None, ""

    # --------------------------------------------------------
    # Generate normal AI-style responses
    # --------------------------------------------------------

    def generate_response(self, text):

        text_lower = text.lower().strip()

        if text_lower in ["hello", "hi", "hey"]:
            return "Hello. I'm Jarvis. How can I help you?"

        if "how are you" in text_lower:
            return "I'm running perfectly and ready to help."

        if "who are you" in text_lower:
            return (
                "I am Jarvis, your personal AI assistant."
            )

        if "thank you" in text_lower or "thanks" in text_lower:
            return "You're welcome."

        if "good morning" in text_lower:
            return "Good morning. Jarvis is online."

        if "good night" in text_lower:
            return "Good night. I'll be here when you need me."

        return (
            "I understand what you're saying. "
            "My generative AI system can handle "
            "more advanced conversations once connected."
        )

    # --------------------------------------------------------
    # Main processing function
    # --------------------------------------------------------

    def process(self, text):

        command, argument = self.understand(text)

        # If it is an executable command
        if command:

            result = execute_command(
                command,
                argument
            )

            if command == "open_youtube":
                response = "Opening YouTube."

            elif command == "open_google":
                response = "Opening Google."

            elif command == "open_gmail":
                response = "Opening Gmail."

            elif command == "open_whatsapp":
                response = "Opening WhatsApp."

            elif command == "google_search":
                response = (
                    f"Searching Google for {argument}."
                )

            elif command == "youtube_search":
                response = (
                    f"Searching YouTube for {argument}."
                )

            elif command == "gmail_search":
                response = (
                    f"Searching Gmail for {argument}."
                )

            elif command == "get_time":
                response = f"The time is {result}."

            elif command == "get_date":
                response = f"Today is {result}."

            elif command == "call":
                response = (
                    f"I received your request to call "
                    f"{argument}. Phone integration is "
                    f"being connected."
                )

            else:
                response = "Command completed."

            self.remember(text, response)

            return response

        # Otherwise treat it as conversation
        response = self.generate_response(text)

        self.remember(text, response)

        return response


# ============================================================
# TEST
# ============================================================

if __name__ == "__main__":

    jarvis = JarvisBrain()

    print("JARVIS 3.2 BRAIN ONLINE")
    print("Type 'exit' to stop.")

    while True:

        user_input = input("You: ")

        if user_input.lower() == "exit":
            print("Jarvis: Goodbye.")
            break

        response = jarvis.process(user_input)

        print("Jarvis:", response)
